R/track.R

```diff
@@ -0,0 +1,403 @@
+  
+  features <- list(
+    angle(tracks),
+    chemotacticIndex(tracks),
+    directionality(tracks),
+    distance(tracks),
+    directionalPersistence(tracks),
+    forwardMigrationIndex(tracks),
+    lifeTime(tracks),
+    meanSquaredDisplacement(tracks,tau = 2),
+    sectorAnalysis(tracks),
+    speed(tracks)) 
+  
+  return(Reduce(function(...) merge(..., all = TRUE, by = grouping_variable), features))
+}
+
+#' Add spatial displacement per frame for each track object
+#'
+#' @param population, data frame storing single cell data 
+#' @param group_variable, column name storing the track label
+#' @return displacement   
+#' @examples 
+#'  data <- tibble::data_frame(
+#'    Metadata_timePoint = c(1:5),
+#'    Location_Center_X = c(1, 2, 3, 4, 5),
+#'    Location_Center_Y = c(1, 1, 1, 1, 1),
+#'    TrackObjects_Label = c(rep(1, 5))
+#'  )
+#'  tracks <- cytominer::displace(data,'TrackObjects_Label')
+#' @importFrom magrittr %>%
+#' @export
+displace <- function(population, group_variable) {
+  dplyr::right_join(
+    population %>%
+      dplyr::select_(.dots = c(group_variable, 'Location_Center_X', 'Location_Center_Y','Metadata_timePoint')) %>%
+      dplyr::mutate(Metadata_timePoint2 =  (Metadata_timePoint - 1) ) %>% 
+      dplyr::filter(Metadata_timePoint2 != -1) %>%
+      dplyr::select(-Metadata_timePoint), 
+    population, by = (.dots = c(group_variable, "Metadata_timePoint2" = "Metadata_timePoint" ))) %>%
+      dplyr::mutate(Track_dX = Location_Center_X.x - Location_Center_X.y) %>% 
+      dplyr::mutate(Track_dY = Location_Center_Y.x - Location_Center_Y.y) %>% 
+      dplyr::select(-Location_Center_X.x, -Location_Center_Y.x) %>%
+      dplyr::rename(Location_Center_X = Location_Center_X.y) %>%
+      dplyr::rename(Location_Center_Y = Location_Center_Y.y) %>% 
+      dplyr::rename(Metadata_timePoint = Metadata_timePoint2) %>%
+      dplyr::mutate(TrackObjects_Distance_Traveled = sqrt(Track_dX^2 + Track_dY^2)) 
+}
+
+#' Add spatial displacement per frame for each track object
+#'
+#' @param tracks data frame with single cell data 
+#' @return displacement   
+#' @examples 
+#'  data <- tibble::data_frame(
+#'    Metadata_timePoint = c(1:5),
+#'    Location_Center_X = c(1, 2, 3, 4, 5),
+#'    Location_Center_Y = c(1, 1, 1, 1, 1),
+#'    TrackObjects_Label = c(rep(1, 5))
+#'  )
+#'  tracks <- cytominer::displace(data,'TrackObjects_Label')
+#'  speed <- cytominer::speed(tracks)
+#' @importFrom magrittr %>%
+#' @export
+speed <- function(tracks) {
+  tracks %>%
+    dplyr::summarize(Track_Length = n(),
+      Track_Speed = sum(TrackObjects_Distance_Traveled, na.rm = TRUE) / (n() - 1), 
+      Track_Speed_max = max(TrackObjects_Distance_Traveled, na.rm = TRUE),
+      Track_Speed_X = sum(Track_dX, na.rm = TRUE) / (n() - 1),
+      Track_Speed_Y = sum(Track_dY, na.rm = TRUE) / (n() - 1)) %>%
+    dplyr::select(-Track_Length)
+  
+}
+
+#' Computer the forward migration index of a track object
```

Comment:
Not sure. The implemented verb is track and the tracks are stored in an on track object. Does that make sense?